feat: extract most globals

hjdivad opened the pull request:
Copies most bins from
<https://github.com/rwjblue/dotfiles/tree/master/binutils/crates/global/src/bin> into a new crate, `global`.

The intended use is:

1. `$PATH` contains something like:

  - "$HOME/src/github/hjdivad/dotfiles/local-packages/crates/global/target/debug"
  - "$HOME/src/github/hjdivad/dotfiles/packages/binutils/crates/global/target/debug"
  - "$HOME/src/github/malleatus/shared_binutils/global/target/debug"

2. dotfiles still has its own `rs/generate-binutils-symlinks.rs`

3. `cargo build && ./target/debug/generate-binutils-symlinks` is ran in
   both dotfiles and shared_binutils

`generate-binutils-symlinks` needs duplication because what it inspects
depends on where it is built.

`rebuild_binutils` wasn't copied because it seems pointless if all the
bins already rely on `latest_bin`.

hjdivad committed:
feat: extract most globals

Copies most bins from
<https://github.com/rwjblue/dotfiles/tree/master/binutils/crates/global/src/bin> into a new crate, `global`.

The intended use is:

1. `$PATH` contains something like:

  - "$HOME/src/github/hjdivad/dotfiles/local-packages/crates/global/target/debug"
  - "$HOME/src/github/hjdivad/dotfiles/packages/binutils/crates/global/target/debug"
  - "$HOME/src/github/malleatus/shared_binutils/global/target/debug"

2. dotfiles still has its own `rs/generate-binutils-symlinks.rs`

3. `cargo build && ./target/debug/generate-binutils-symlinks` is ran in
   both dotfiles and shared_binutils

`generate-binutils-symlinks` needs duplication because what it inspects
depends on where it is built.

`rebuild_binutils` wasn't copied because it seems pointless if all the
bins already rely on `latest_bin`.

rwjblue commented on test_utils/src/lib.rs:
```diff
@@ -17,8 +17,8 @@ pub struct TestEnvironment {
 impl Drop for TestEnvironment {
     fn drop(&mut self) {
         match &self.original_home {
-            Some(home) => unsafe { env::set_var("HOME", home) },
-            None => unsafe { env::remove_var("HOME") },
```
Why remove the unsafe here? Without the unsafe I get rust-analyzer warnings locally

hjdivad replied:
I only removed it because clippy was complaining about a useless unsafe.

rwjblue commented:
> Copies most bins from
https://github.com/rwjblue/dotfiles/tree/master/binutils/crates/global/src/bin 

Which SHA did you start from? (just to make it easier for me to port some of my local changes on top)

rwjblue commented:
I'm fairly sure the tmux tests are failing because I've never ran them _outside_ of a tmux window locally (but in CI here they are)

rwjblue committed:
Ignore WIP test (test_linked_crates)

rwjblue committed:
Migrate to using debug snapshots throughout

The YAML sanpshots were a leftover from when config was yaml

rwjblue committed:
Bring back unsafe (until I grok why it was removed)

rwjblue committed:
Make tests stable WRT: tmux status

rwjblue committed:
Make sure to handle std::env::current_exe differences per platform

On macOS the `std::env::current_exe()` path typically does not get
resolved (so we can assert that the current_exe is actually within the
symlinked path), but on Linux the `current_exe()` result is usually
canonicalized (such that the symlink is no longer part of the path, but
it is instead the "real" bin path).

The rest of our logic doesn't actually care if we are seeing a symlink
or not, we are just trying to confirm the executable works.
